### ISSUE TITLE
false positive in `object_usage_linter()` with `substitute()` in function

### ISSUE THREAD

**Issue by jmbarbone** [r-lib/lintr#2255]
Something funny seems to be happening with using `substitute()` when assigning a `function` definition:

``` r
library(lintr)
packageVersion("lintr")
#> [1] '3.1.0.9000'

code_ok1 <- "
substitute({
  a <- 1
  a
})
"

code_ok2 <- "
function() {
  substitute({
    a <- 1
    a
  })
}
"

code_bad <- "
foo <- function() {
  substitute({
    a <- 1
    a
  })
}
"

lint(text = code_ok1, linters = object_usage_linter())
lint(text = code_ok2, linters = object_usage_linter())
lint(text = code_bad, linters = object_usage_linter())
#> <text>:4:5: warning: [object_usage_linter] local variable 'a' assigned but may not be used
#>     a <- 1
#>     ^
```

<sup>Created on 2023-11-06 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by AshesITR** [comment]
IINM the first two are not linted at all by `object_usage_linter()`.
The last one appears to be a bug in `codetools::checkUsage()`.

**Comment by travis-leith** [comment]
Here is another case where `object_usage_linter` gives a false positive

```R

object_linter_fail <- function(){
  a <- 1
  
  b <- case_when(
    FALSE ~ "",
    TRUE ~ paste0("", a, "")
  )
  
  b
}
```

It claims that a is not used.

**Comment by AshesITR** [comment]
`dplyr::case_when()` uses NSE, making this hard for `codetools::checkUsage()` to detect.
Syntactically, formulae such as `TRUE ~ paste0("", a, "")` are not evaluated and thus their contained code not "used".

**Comment by travis-leith** [comment]
@AshesITR is it worth raising an issue in the code tools repo?
https://gitlab.com/luke-tierney/codetools/-/issues
or is this already a well known limitation?

**Comment by AshesITR** [comment]
Feel free to raise an issue. My experience with issues regarding codetools is they probably won't be fixed unless you provide a patch.

**Comment by travis-leith** [comment]
I tried to raise an issue but gitlab wants a credit card number to create an account, and I am not willing to provide it.

**Comment by AshesITR** [comment]
I had no problem logging into GitLab using GitHub as an identity provider. But your mileage may vary.
